README.md

# Virtual Background Bodypix :computer:

## Description :page_facing_up:

Using getusermedia as a video input, we offer the effects blur background, blur bodyParts, virtualBackground, grayScale having bodyPix as a backbone!
Use bodypix in 3 lines!, making easy for non developers and developers to use bodypix with multiple effects.

## Install

Can install it via npm for use in a TypeScript / ES6 project.

```sh
$ npm i tribe_lib
```

### How To Use

After installing we simply use our three magic lines, to get a MediaStream Object that you can use for wherever you want.

```js
//Common Javascript
const virtualBackgroundBodypixLite = require('tribe_lib');

//ES6
import { VideoTracking, Prediction } from 'tribe_lib';

//How to Use
const Tracking = new VideoTracking(
  type_model_architecture,
  effect_config_type,
  type_of_device
  width:number, 
  height:number, 
  device_id_str?:string
);
Tracking.predictionModel.loop_(type_prediciton, configEffect);
const MediaStream = Tracking.predictionModel.canvas_mediaStream();
```
### How To Reuse Object
if you want to reuse the VideoTracking object after initialazing it and starting a Loop  do the following.
```js
const Tracking = new VideoTracking(
  type_model_architecture,
  effect_config_type,
  type_of_device
  width:number, 
  height:number, 
  device_id_str?:string
);
Tracking.predictionModel.loop_(type_prediciton, configEffect);

//Stop Actual Loop
Tracking.predicitionModel.stopAnimationLoop();

//start another desired Effect
Tracking.predictionModel.loop_(type_prediction_2, configEffect_2);
```

### Models

We can define 5 categories **_Ultra low_**, **_Low_**, **_Medium_**, **_High_**, **_Ultra_** for model configuration

| Categories | Description                                                         | Option |
| ---------- | ------------------------------------------------------------------- | ------ |
| Ultra low  | Recommended for low-end mobile devices                              | 0      |
| Low        | Recommended for mid-end mobile devices                              | 1      |
| Medium     | Recommended for computers with at least one intel pentium processor | 2      |
| High       | Recommended for computers with at least one intel core i3 processor | 3 

```js
const model_config_ultra_low = {
  architecture: 'MobileNetV1',
  outputStride: 16,
  multiplier: 0.5,
  quantBytes: 2,
};
const model_config_low = {
  architecture: 'MobileNetV1',
  outputStride: 16,
  multiplier: 0.75,
  quantBytes: 2,
};
const model_config_medium = {
  architecture: 'MobileNetV1',
  outputStride: 16,
  multiplier: 1,
  quantBytes: 2,
};
const model_config_high = {
  architecture: 'MobileNetV1',
  outputStride: 8,
  multiplier: 1,
  quantBytes: 2,
};
const model_config_ultra = {
  architecture: 'ResNet50',
  outputStride: 16,
  quantBytes: 2,
};
```

### Level Of Prediction

We can define the level of prediction
|Categories | Resolution |Option
|--|--|--|
| Low | Minimum|0
| Medium| Average|1
| High| High|2
| ultra| ultra|3

```js
const effect_config_precission_low = {
  flipHorizontal: false,
  internalResolution: 'low',
  segmentationThreshold: 0.7,
};
const effect_config_precission_mid = {
  flipHorizontal: false,
  internalResolution: 'medium',
  segmentationThreshold: 0.7,
};
const effect_config_precission_high = {
  flipHorizontal: false,
  internalResolution: 'high',
  segmentationThreshold: 0.7,
};
const effect_config_precission_high = {
  flipHorizontal: false,
  internalResolution: 'ultra',
  segmentationThreshold: 0.7,
};
```

### Type Of Devices

We can define the dimensions of the video, in addition to which camera to request in mobile.

```js
const mobile_front_camera = {
  audio: false,
  video: { facingMode: 'user', width: width, height: height },
};
const mobile_rear_camera = {
  audio: false,
  video: { facingMode: { exact: 'environment' }, width: width, height: height },
};
const desktop_selected_camera_device = {
  audio: false,
  video: {
    deviceId: '0faf4c1dc3b35ff09df6a31...',
    width: width,
    height: height,
  },
};
const desktop_select_camera_default = {
  audio: false,
  video: { width: width, height: height },
};
```

### Config Effects

These are the effects, the magic of this library, you can define what effect use:

#### Grey scale

![Image](https://i.imgur.com/d8Xs7i0.png[/img])

![Image](https://i.imgur.com/OgyT3Sl.png[/img])

```js
Tracking.predictionModel.loop_(3);
const MediaStream = Tracking.predictionModel.canvas_mediStream();
Tracking.predictionModel.stopAnimationLoop();/*To Stop Loop*/
```

#### Blur

![Image](https://i.imgur.com/xkAT4pf.png[/img])

```js
Tracking.predictionModel.loop_(1, config_effect_bokek);
const MediaStream = Tracking.predictionModel.canvas_mediaStream();
Tracking.predictionModel.stopAnimationLoop();/*To Stop Loop*/
```

#### Virtual Background

![Image](https://i.imgur.com/m42kzl8.png[/img])

```js
Tracking.predictionModel.loop_(2, config_virtual_background);
const MediaStream = Tracking.predictionModel.canvas_mediaStream();
Tracking.predictionModel.stopAnimationLoop();/*To Stop Loop*/
```

#### Blur Body Parts

![Image](https://i.imgur.com/SKfLbIB.png[/img])

```js
const parts: partsbody = {
  left_face: 0,
  torso_front: 12,
  right_face: 1,
  torso_back: 13,
  left_upper_arm_front: 2,
  left_upper_leg_front: 14,
  left_upper_arm_back: 3,
  left_upper_leg_back: 15,
  right_upper_arm_front: 4,
  right_upper_leg_front: 16,
  right_upper_arm_back: 5,
  right_upper_leg_back: 17,
  left_lower_arm_front: 8,
  left_lower_leg_front: 18,
  left_lower_arm_back: 7,
  left_lower_leg_back: 19,
  right_lower_arm_front: 8,
  right_lower_leg_front: 20,
  right_lower_arm_back: 9,
  right_lower_leg_back: 21,
  left_hand: 10,
  left_foot: 22,
  right_hand: 11,
  right_foot: 23,
};
const config_blur_body_part = { backgroundBlurAmount: 30, edgeBlurAmount: 2.1, faceBodyPartIdsToBlur: [0, 1] };
Tracking.predictionModel.loop_(4, config_blur_body_part);
const MediaStream = Tracking.predictionModel.canvas_mediaStream();
Tracking.predictionModel.stopAnimationLoop();/*To Stop Loop*/
```

```js
const config_effect_bokek = { backgroundBlurAmount: 20, edgeBlurAmount: 10 };
const config_virtual_background = {
  backgroundBlurAmount: 1,
  edgeBlurAmount: 2.1,
  URL: 'base64',
};
const config_greyScale = {};
const config_blur_body_part = {
  backgroundBlurAmount: 30,
  edgeBlurAmount: 2.1,
  faceBodyPartIdsToBlur: [0, 1],
};
```
## More information about the model
**Repo BodyPix:** [Github](https://github.com/tensorflow/tfjs-models/tree/master/body-pix)
## Authors :black_nib:

**Hector Lopez** [Github](https://github.com/hectorlopezv)

**Hugo Fernel** [Github](https://github.com/daviddlhz)

**Jhonathan Angarita** [Github](https://github.com/JhonathanAlejandro01)

**David De La Hoz** [Github](https://github.com/daviddlhz)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.23](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.22...v1.0.23) (2020-11-04)

### [1.0.22](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.21...v1.0.22) (2020-11-04)

### [1.0.21](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.20...v1.0.21) (2020-10-30)

### [1.0.20](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.19...v1.0.20) (2020-10-30)

### [1.0.19](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.18...v1.0.19) (2020-10-30)

### [1.0.18](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.17...v1.0.18) (2020-10-30)

### [1.0.17](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.16...v1.0.17) (2020-10-30)

### [1.0.16](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.15...v1.0.16) (2020-10-30)

### [1.0.14](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.15...v1.0.14) (2020-10-30)

### [1.0.13](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.15...v1.0.13) (2020-10-30)

### [1.0.12](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.11...v1.0.12) (2020-10-29)

### [1.0.11](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.10...v1.0.11) (2020-10-29)

### [1.0.10](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.9...v1.0.10) (2020-10-29)

### [1.0.9](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.8...v1.0.9) (2020-10-29)

### [1.0.8](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.7...v1.0.8) (2020-10-29)

### [1.0.7](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.6...v1.0.7) (2020-10-29)

### [1.0.6](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.5...v1.0.6) (2020-10-29)

### [1.0.5](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.4...v1.0.5) (2020-10-29)

### [1.0.4](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.3...v1.0.4) (2020-10-29)

### [1.0.3](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.2...v1.0.3) (2020-10-29)

### [1.0.2](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.1...v1.0.2) (2020-10-29)

### [1.0.1](https://github.com/hectorlopezv/tribe_lib/compare/v1.0.0...v1.0.1) (2020-10-27)

## 1.0.0 (2020-10-27)


src/lib/wrapper.ts
/*
 * Tribe Lib Package
 * Copyright 2020
 * Author: Kayman Lab
 * All Rights Reserved.
 * Use, reproduction, distribution, and modification of this code is subject to the terms and
 * conditions of the MIT license, available at http://www.opensource.org/licenses/mit-license.php
 *
 */


import * as bodyPix from '@tensorflow-models/body-pix';


export class VideoTracking {

   /**
    * @desc Class that contain all about how get a MediaStream.
    * 
    * @param {Array} model_architeture_options Differents options of models to use.
    * @param {Array} effect_config_precission Types of segmentation.
    * @param {Array} type_of_device Define the dimensions of the video, in addition to which camera to request in mobile.
    * @param {Object} selection_model_option Save the model selected.
    * @param {Object} selection_effect_option Save the effect selected.
    * @param {Object} selection_type_device Save the type of device selected.
    * @param {Object} canvasElemenet contains a canvas element.
    * @param {Object} VideoElement contains a video element.
    * @param {Object} model contains the model load.
    * @param {Object} Video contains a Promesis of a video tag.
    * @param {String} deviceId contains the id of the video device.
    * @param {Number} frameRate contains frame rate of the segmentation.
    * @param {Object} video_stream contains a MediaStream of video tag.
    * @param {Object} predictionModel Save the prediction.
    *
    */
    

    model_architeture_options: Array<ModelConfig>;
    effect_config_precission: Array<effect_config>;
    type_of_device: Array<typeDeviceConfig>;
    selection_model_option: ModelConfig;
    selection_effect_option: effect_config;
    selection_type_device: typeDeviceConfig;
    canvasElemenet: CanvasElement; 
    VideoElement: HTMLVideoElement;
    model:Promise<unknown>;
    Video: Promise<HTMLVideoElement>;
    deviceId: string;
    frameRate: number;
    video_stream: MediaStream;
    predictionModel: Prediction;

    constructor(model_config_number: number, config_prediction_number:number, type_device_number:number, width:number, height:number, device_id_str?:string  ){
  
        this.model_architeture_options = [{architecture: "MobileNetV1", outputStride: 16, multiplier: 0.5, quantBytes: 2,},//Use in load model
            { architecture: 'MobileNetV1', outputStride: 16, multiplier: 0.75, quantBytes: 2},
            { architecture: 'MobileNetV1', outputStride: 16, multiplier: 1, quantBytes: 2}, 
            { architecture: 'MobileNetV1', outputStride: 8, multiplier: 1, quantBytes: 2},
            { architecture: 'ResNet50', outputStride: 16, quantBytes: 2}];

        this.effect_config_precission = [{  flipHorizontal: false, internalResolution: 'low', segmentationThreshold: 0.7},//use in Prediction
            {  flipHorizontal: false, internalResolution: 'medium', segmentationThreshold: 0.7},
            {  flipHorizontal: false, internalResolution: 'high', segmentationThreshold: 0.7},
            {  flipHorizontal: false, internalResolution: 'ultra', segmentationThreshold: 0.7}];

        this.type_of_device = [ { audio: false, video: { facingMode: "user", width: width, height: height }},//use in load_Video
            { audio: false, video: { facingMode: { exact: "environment" }, width: width, height: height }},
            { audio: false, video: {deviceId: device_id_str,  width: width, height: height }},
            { audio: false, video: { width:  width, height: height }}];

        this.selection_model_option =  this.model_architeture_options[model_config_number];
        this.selection_effect_option = this.effect_config_precission[config_prediction_number];
        this.selection_type_device = this.type_of_device[type_device_number];
        this.model = this._load_model(this.selection_model_option);
        this.Video =  this.load_Video_stream(this.selection_type_device);
        this.canvasElemenet = this.createCanvas(width, height);
        this.VideoElement = this.createVideo(width, height);
        this.predictionModel = new Prediction (this.model, this.Video, this.canvasElemenet, this.selection_effect_option, width, height);

    }

    createCanvas(width:number, height:number):CanvasElement {
       /**
        * @desc Create a canva element
        * @param {number} width width of the canva.
        * @param {number} height height of the canva.
        *
        * @return canva element
        */

        const canvas: CanvasElement = <CanvasElement>document.createElement('canvas');
        canvas.setAttribute('width', String(width) );
        canvas.setAttribute('height', String(height) );
        return canvas
      }

    createVideo(width:number, height:number): HTMLVideoElement {
       /**
        * @desc Create a video element
        * @param {number} width width of the video.
        * @param {number} height height of the video.
        *
        * @return video element
        */

        const video: HTMLVideoElement= document.createElement('video');
        video.setAttribute('autoplay','false');
        video.setAttribute('width', String(width) );
        video.setAttribute('height', String(height) );
        video.style.display = 'none';
        return video;
    }

    _load_model(config_model: ModelConfig): Promise<unknown> {
       /**
        * @desc Load the model with the configuration.
        * @param {Object} config_model Configuration of model.
        *
        * @return Model loaded
        */

        return  bodyPix.load(config_model);
    }

    async load_Video_stream(config_constrains:typeDeviceConfig):Promise<HTMLVideoElement>{
       /**
        * @desc Obtain the stream of the video element.
        * @param {Object} config_constrains Configuration of model.
        *
        * @return a promise that load the video stream
        */
     
        const stream: MediaStream = await navigator.mediaDevices.getUserMedia(config_constrains);

        this.VideoElement.srcObject = stream;
        this.video_stream = stream;
        return this.PromiseCreator();
    }

    PromiseCreator (): Promise<HTMLVideoElement>{ /* Return */
    /**
     * @desc Set width and height in the video element
     *
     * @return a promise VideoElement when all data is loaded and ready  with certains dimensions
     */
        return new Promise((resolve) => {
            this.VideoElement.onloadedmetadata = () => {
                this.VideoElement.width = this.VideoElement.videoWidth;
                this.VideoElement.height = this.VideoElement.videoHeight;
            resolve(this.VideoElement);
          };
        });
      }
  
    static addVideo(HTMLelement:HTMLElement, videoElement:CanvasElement|HTMLVideoElement) {
        HTMLelement.appendChild(videoElement);
      }
    
}

export class Prediction {

    /**
    * @desc Class that contain all about how get a prediction.
    * 
    * @param {Promise} loaded_model contains a model loaded.
    * @param {Promise} videoMediaStream contains a MediaStream of video element.
    * @param {Boolean} stop Stop the loop of segmentation.
    * @param {Object} canvasElement constains the canvas element.
    * @param {Object} canvas_stream contains a MediaStream of canvas element.
    * @param {Object} loaded_video Contains the video.
    * @param {Object} height height the video and canva element.
    * @param {Object} width width the video and canva element.
    * @param {String} deviceId Contains the id of the video device.
    * @param {Number} frameRate Contains frame rate of the segmentation.
    * @param {Object} selection_effect_option The option that contains the effect selected.
    * @param {Object} model_prediction Save the prediction.
    */


    loaded_model: Promise<unknown>;
    videoMediaStream: Promise<HTMLVideoElement>;
    stop: boolean;
    canvasElement: CanvasElement;
    canvas_stream: MediaStream;
    loaded_video: HTMLVideoElement|undefined;
    height: number;
    width: number;
    deviceId: string;
    frameRate: number;
    selection_effect_option: effect_config;
    model_prediction: any;

    constructor( loaded_model:Promise<unknown>, videoMediaStream:Promise<HTMLVideoElement>, canvasElemenet:CanvasElement, selection_effect_option:effect_config, width:number, height:number){
         this.loaded_model = loaded_model;
         this.videoMediaStream = videoMediaStream;
         this.stop = false;
         this.canvasElement = canvasElemenet;
         this.selection_effect_option = selection_effect_option;
         this.width = width;
         this.height = height;
     }
     
    static addHTMLCanvas_Video(parentNode:HTMLElement, nodeToAdd:CanvasElement|HTMLVideoElement){
        parentNode.appendChild(nodeToAdd);
    }

    canvas_mediaStream(fps:number):MediaStream{  
        /**
        * @desc Get the MediaStream of the canva
        * @param {Number} fps Frames per second
        *
        * @return the stream of canva
        */

        const ctx:CanvasRenderingContext2D = this.canvasElement.getContext("2d");
        console.log(ctx);
        const stream: MediaStream= this.canvasElement.captureStream(fps);
        return stream;
    }

     async videoImageData (width:number, height:number, videoElement:HTMLVideoElement):Promise<ImageData>{
        /**
        * @desc Get video data
        * @param {Number} width width the video
        * @param {Number} height height the video
        * @param {Object} videoElement the video element
        *
        * @return the data of video
        */
    
        const canvas:CanvasElement = <CanvasElement>document.createElement('canvas');
        canvas.setAttribute('width', String(width) );
        canvas.setAttribute('height', String(height));
        
        const context:CanvasRenderingContext2D = canvas.getContext('2d');
        const img = videoElement;
        
        context.drawImage(img, 0, 0 );
        const theData:ImageData = context.getImageData(0, 0, width, height);
        return theData;
    }

    async getImageData(width:number, height:number, base64:string): Promise<ImageData>{
        /**
        * @desc Get image the video element
        * @param {Number} width width the video
        * @param {Number} height height the video
        * @param {String} base64 --
        *
        * @return the data of image
        */

        const canvas:CanvasElement = <CanvasElement>document.createElement('canvas');
        canvas.width = width;
        canvas.height = height;
        const ctx:CanvasRenderingContext2D = canvas.getContext('2d');

        const img = new Image();
        img.crossOrigin = '';
        img.src = base64;
        
        ctx.drawImage(img, 0, 0, img.width, img.height, 0,0, width, height);
        return ctx.getImageData(0, 0, width, height);   
    }

    stopAnimationLoop():boolean{
        /**
        * @desc Stop the segmentation in the canva
        *
        * @return The option of stop the segmentation.
        */

        this.stop = true;
        return this.stop;
    }

    async make_prediction_load():Promise<void>{ 
        /**
        * @desc Load the model and Get the prediction
        *
        */

        if (typeof this.loaded_video === 'undefined'){
            this.loaded_video = await this.videoMediaStream;
            this.model_prediction = await this.loaded_model;
        }
    }

    async effect_blur_background(canvasElement:CanvasElement, image:HTMLVideoElement, personSegmentation:SemanticPersonSegmentation ,  config:effect_config):Promise<void> {
        /**
        * @desc Get the effect blur
        *
        * @param {Object} canvasElement the canvas element.
        * @param {Object} image image data of video element.
        * @param {Object} personSegmentation model segmentation.
        * @param {Object} config configuration of effect blur.
        *
        * @return The prediction or segmentation of the person
        */

        const {backgroundBlurAmount, edgeBlurAmount, flipHorizontal} = config; 
        await bodyPix.drawBokehEffect(canvasElement, image, personSegmentation, backgroundBlurAmount, edgeBlurAmount, flipHorizontal);
    }

    async virtualBackground_(prediction:SemanticPersonSegmentation, canvasElement: CanvasElement, videoElement:HTMLVideoElement, config:effect_config, base64_img:string){
        /**
        * @desc Config the virtual background
        *
        * @param {Object} prediction Prediction of segmentation.
        * @param {Object} canvasElement Canvas element.
        * @param {Object} videoElement Video element.
        * @param {Object} config Configuration of Virtual Background.
        * @param {String} base64_img Image information.
        *
        */

        const canvas:CanvasElement = canvasElement;
        const newImg:ImageData = canvas.getContext('2d').createImageData(this.width, this.height)
        const newImgData = newImg.data;

        //Prediction
        const {data:map} = prediction;
        const pixelLength:number = map.length;
        
        //Video Data
        const { data: videoData } = await this.videoImageData(this.width, this.height, videoElement);
    
        //ImageData
        const {data: imgData} = await this.getImageData(this.width, this.height, base64_img);

        for (let i = 0; i < pixelLength; i++) {

            const [r, g, b, a] = [imgData[i*4], imgData[i*4 + 1], imgData[i*4 + 2], imgData[i*4 + 3]];
            [
                newImgData[i*4], 
                newImgData[i*4 +1],
                newImgData[i*4 + 2], 
                newImgData[i*4 + 3]
            ] = !map[i]   ? [r, g, b, a] : [
                                            videoData[i*4], 
                                            videoData[i*4 +1], 
                                            videoData[i*4 + 2], 
                                            0]
        }
        this.canvasElement.getContext('2d').clearRect(0,0,this.width, this.height);
        const {  backgroundBlurAmount, edgeBlurAmount, flipHorizontal } = config;
        bodyPix.drawMask(canvas, videoElement, newImg, backgroundBlurAmount, edgeBlurAmount, flipHorizontal);    
    }
    
    async virtual_background(canvasElement: CanvasElement, videoElement:HTMLVideoElement, personSegmentation:SemanticPersonSegmentation,  config:effect_config, base64_img:string){
        /**
        * @desc Get the virtual background
        *
        * @param {Object} canvasElement Canvas element.
        * @param {Object} videoElement Video element.
        * @param {Object} personSegmentation Segmentation of model.
        * @param {Object} config Configuration of Virtual Background.
        * @param {String} base64_img Image information.
        *
        */
        await this.virtualBackground_(personSegmentation, canvasElement, videoElement, config, base64_img);
    }
 
    async blurBodyPart_(canvasElement: CanvasElement, videoElement:HTMLVideoElement, personSegmentationParts:SemanticPartSegmentation,  config:effect_config) {
        /**
        * @desc Get the Blur body parts effect
        *
        * @param {Object} canvasElement Canvas element.
        * @param {Object} videoElement Video element.
        * @param {Object} personSegmentation Segmentation of model.
        * @param {Object} config Configuration of Virtual Background.
        *
        */
        
        /*Reference of Body Parts*/
        const {backgroundBlurAmount, edgeBlurAmount, flipHorizontal, faceBodyPartIdsToBlur} = config;
        await bodyPix.blurBodyPart(canvasElement, videoElement, personSegmentationParts, faceBodyPartIdsToBlur,
            backgroundBlurAmount, edgeBlurAmount, flipHorizontal);
    }

    async grayScale(canvasElement: CanvasElement, videoElement:HTMLVideoElement, personSegmentation:SemanticPersonSegmentation){
        /**
        * @desc Get gray scale effect
        *
        * @param {Object} canvasElement Canvas element.
        * @param {Object} videoElement Video element.
        * @param {Object} personSegmentation Segmentation of model.
        *
        */

        const {data:map} = personSegmentation;
        // Extracting video data
        const { data:imgData } = await this.videoImageData(this.width, this.height, videoElement);

        //New canvas
        const canvas:CanvasElement = canvasElement;

        /*Clean Canvas*/
        const newImg:ImageData = canvas.getContext('2d').createImageData(canvas.width, canvas.height);
        const newImgData = newImg.data;

        //[r0, g0, b0, a0, r1, g1, b1, a1, ..., rn, gn, bn, an]
        for (let i = 0; i < map.length; i++) {

            const [r, g, b, a]: number[] = [imgData[i*4], imgData[i*4+1], imgData[i*4+2], imgData[i*4+3]];
            // GrayScale Effect
            const gray = ((0.3 * r) + (0.59 * g) + (0.11 * b));
            [
                newImgData[i*4],
                newImgData[i*4+1],
                newImgData[i*4+2],
                newImgData[i*4+3]
            ] = !map[i] ? [gray, gray, gray, 255] : [r, g, b, a];
        }
        canvas.getContext('2d').putImageData(newImg, 0, 0);/*Paint the canvas*/
    }

    async loop_(type_prediciton:number, config_:effect_config){
        /**
        * @desc Loop animation that use segmentation
        *
        * @param {Number} type_prediciton Type of prediction and segmentation.
        * @param {Object} config_ Conguration of effect.
        *
        */

        await this.make_prediction_load();
        const loaded_video:HTMLVideoElement = this.loaded_video;
        const model_prediction = this.model_prediction;
        const config = {...config_, ...this.selection_effect_option};
        const {base64_img } = config;
    
        const loopping = async () =>{/*Loop for animation*/
            if(this.stop){
                /*if the is  a loop running stop it*/
                this.stop = false;/*making the object reusable*/
                return;
            }

            if (type_prediciton === 1){/*Blur Background - - PersonSegmentation*/
               
                const prediction_frame = await model_prediction.segmentPerson(loaded_video, config);
                this.effect_blur_background(this.canvasElement, this.loaded_video, prediction_frame,  config);
            }
            
            else if (type_prediciton === 2){//Virtual Background - PersonSegmentation
                
                const prediction_frame:SemanticPersonSegmentation = await model_prediction.segmentPerson(loaded_video, config);     
                this.virtual_background(this.canvasElement, this.loaded_video, prediction_frame,  config, base64_img);
            }
            else if(type_prediciton === 3){//Gray SCale - PersonSegmentation
                
                const prediction_frame:SemanticPersonSegmentation = await model_prediction.segmentPerson(loaded_video, config);     
                this.grayScale(this.canvasElement, this.loaded_video, prediction_frame);
            }
            
            else if(type_prediciton === 4){//Blur Body PARTS - PersonSegmentationPARTS
                const prediction_frameParts:SemanticPartSegmentation = await model_prediction.segmentPersonParts(loaded_video, config); 
                this.blurBodyPart_(this.canvasElement, this.loaded_video, prediction_frameParts, config);
            }
            
            window.requestAnimationFrame(loopping);/*Recursive call*/
       }
    loopping();/*Closure*/
    } 
 }


src/types/example.d.ts
/**
 * If you import a dependency which does not include its own type definitions,
 * TypeScript will try to find a definition for it by following the `typeRoots`
 * compiler option in tsconfig.json. For this project, we've configured it to
 * fall back to this folder if nothing is found in node_modules/@types.
 *
 * Often, you can install the DefinitelyTyped
 * (https://github.com/DefinitelyTyped/DefinitelyTyped) type definition for the
 * dependency in question. However, if no one has yet contributed definitions
 * for the package, you may want to declare your own. (If you're using the
 * `noImplicitAny` compiler options, you'll be required to declare it.)
 *
 * This is an example type definition which allows import from `module-name`,
 * e.g.:
 * ```ts
 * import something from 'module-name';
 * something();
 * ```
 */
declare module 'module-name' {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const whatever: any;
  export = whatever;
}


type BodyPixInternalResolution = number|'low'|'medium'|'high'|'full';
type BodyPixOutputStride = 32|16|8;
type BodyPixArchitecture = 'ResNet50'|'MobileNetV1';
type BodyPixQuantBytes = 1|2|4;
type BodyPixMultiplier = 1.0|0.75|0.50;
interface ModelConfig  {
  architecture: BodyPixArchitecture;
  outputStride: BodyPixOutputStride;
  multiplier?: BodyPixMultiplier;
  modelUrl?: string;
  quantBytes?: BodyPixQuantBytes;
}


interface Vector2D  {
  y: number,
  x: number
}
interface Pose  {
  keypoints: Keypoint[],
  score: number,
}
interface CanvasElement extends HTMLCanvasElement { 
  captureStream(frameRate?: number): MediaStream;          
 }
interface Keypoint {
  score: number,
  position: Vector2D,
  part: string
}
interface partsbody {
  'left_face':number;
  'torso_front': number; 
  'right_face':number;
  'torso_back':number;
  'left_upper_arm_front':number; 
  'left_upper_leg_front':number; 
  'left_upper_arm_back':number;            
  'left_upper_leg_back':number;
  'right_upper_arm_front': number; 
  'right_upper_leg_front':number;                   
  'right_upper_arm_back':number; 
  'right_upper_leg_back':number; 
  'left_lower_arm_front':number;                 
  'left_lower_leg_front':number; 
  'left_lower_arm_back':number; 
  'left_lower_leg_back':number;               
  'right_lower_arm_front':number; 
  'right_lower_leg_front':number; 
  'right_lower_arm_back':number;                     
  'right_lower_leg_back':number;
  'left_hand': number;
  'left_foot': number; 
  'right_hand':number;             
  'right_foot': number;
}
interface SemanticPartSegmentation {
  data: Int32Array;
  width: number;
  height: number;
  allPoses: Pose[];
}

interface SemanticPersonSegmentation  {
  data: Uint8Array;
  width: number;
  height: number;
  allPoses: Pose[];
}

interface PartSegmentation  {
  data: Int32Array;
  width: number;
  height: number;
  pose: Pose;
}

interface PersonSegmentation  {
  data: Uint8Array;
  width: number;
  height: number;
  pose: Pose;
}

interface typeDeviceConfig_video {
  width: number;
  height: number;
  deviceId?: string;
  facingMode?: unknown;
}

interface typeDeviceConfig {
  audio: boolean;
  video: typeDeviceConfig_video 
}

interface testing {
  x: number; 

}

interface effect_config {
  backgroundBlurAmount?: number;
  edgeBlurAmount?: number;
  base64_img?: string;
  faceBodyPartIdsToBlur?: number[];
  flipHorizontal?: boolean;
  internalResolution?: string;
  segmentationThreshold?: number;
}

